lib/diskoverflow/slice.go

```diff
@@ -0,0 +1,132 @@
+// Copyright (C) 2018 The Syncthing Authors.
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+package diskoverflow
+
+import (
+	"fmt"
+)
+
+type Slice interface {
+	Append(v Value)
+	NewIterator() Iterator
+	NewReverseIterator() Iterator
+	Bytes() int
+	Items() int
+	SetOverflowBytes(bytes int)
+	Close()
+}
+
+type slice struct {
+	commonSlice
+	base
+}
+
+type commonSlice interface {
+	common
+	append(v Value)
+	newIterator(p iteratorParent, reverse bool) Iterator
+}
+
+// NewSorted creates a slice like container, spilling to disk at location.
+// All items added to this instance must be of the same type as v.
+func NewSlice(location string) Slice {
+	o := &slice{base: newBase(location)}
+	o.commonSlice = &memorySlice{}
+	return o
+}
+
+func (o *slice) Append(v Value) {
+	if o.iterating {
+		panic(concurrencyMsg)
+	}
+	if o.startSpilling(o.Bytes() + v.ProtoSize()) {
+		d, err := v.Marshal()
+		errPanic(err)
+		ds := &diskSlice{newDiskSorted(o.location)}
+		it := o.NewIterator()
+		for it.Next() {
+			v.Reset()
+			it.Value(v)
+			ds.append(v)
+		}
+		it.Release()
+		o.commonSlice.Close()
+		o.commonSlice = ds
+		o.spilling = true
+		v.Reset()
+		errPanic(v.Unmarshal(d))
+	}
+	o.append(v)
+}
+
+func (o *slice) released() {
+	o.iterating = false
+}
+
+func (o *slice) NewIterator() Iterator {
+	return o.newIterator(false)
+}
+
+func (o *slice) NewReverseIterator() Iterator {
+	return o.newIterator(true)
+}
+
+func (o *slice) newIterator(reverse bool) Iterator {
+	if o.iterating {
+		panic(concurrencyMsg)
+	}
```

Comment:
Why is it ok to have two iterators on a map, but not on a slice?

Reply:
It being ok on map is an oversight. Whether this restriction is needed is another topic: Two iterators should indeed not be a problem. What I wanted to prevent is modifying the object while iterating, just because that either requires brains to do it right or data copying to be save. As we don't need modification while iteration, I just took the easy path to forbid it.